File: ios/Selko/Features/Integrations/Models/Integration.swift
//
//  Integration.swift
//  Selko
//

import Foundation

enum IntegrationProvider: String, Codable, Sendable {
    case gmail
    case outlook
    case googlePhotos = "google_photos"
    case googleCalendar = "google_calendar"
}

enum IntegrationStatus: String, Codable, Sendable {
    case active
    case expired
    case revoked
    case error
}

enum IntegrationRecoveryStatus: String, Codable, Sendable {
    case pending
    case processing
    case waiting
    case completed
    case completedWithErrors = "completed_with_errors"
    case failed
    case superseded
}

/// Latest Google Calendar reconnect recovery generation.
///
/// Mirrors `public.integration_recoveries`. RLS exposes only the current
/// user's own rows. Email reconnects need no recovery record (they resume
/// from provider cursors), so only `google_calendar` produces one.
struct IntegrationRecovery: Identifiable, Codable, Sendable, Equatable {
    let id: UUID
    let integrationId: UUID
    let userId: UUID
    let provider: IntegrationProvider
    let status: IntegrationRecoveryStatus
    let discoveredCount: Int?
    let completedCount: Int?
    let remainingCount: Int?
    let errorDetail: String?
    let requestedAt: Date?

    enum CodingKeys: String, CodingKey {
        case id
        case integrationId = "integration_id"
        case userId = "user_id"
        case provider
        case status
        case discoveredCount = "discovered_count"
        case completedCount = "completed_count"
        case remainingCount = "remaining_count"
        case errorDetail = "error_detail"
        case requestedAt = "requested_at"
    }

    var isActive: Bool {
        status == .pending || status == .processing || status == .waiting
    }

    var needingAttentionCount: Int {
        max(0, (discoveredCount ?? 0) - (completedCount ?? 0))
    }
}

struct Integration: Identifiable, Codable, Sendable, Equatable {
    let id: UUID
    let userId: UUID
    let provider: IntegrationProvider
    let status: IntegrationStatus
    let providerEmail: String?
    let scopes: [String]
    let lastSyncAt: Date?
    let createdAt: Date?
    let updatedAt: Date?

    enum CodingKeys: String, CodingKey {
        case id
        case userId = "user_id"
        case provider
        case status
        case providerEmail = "provider_email"
        case scopes
        case lastSyncAt = "last_sync_at"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }

    var isActive: Bool {
        status == .active
    }
}

extension Integration {
    static var mock: Integration {
        Integration(
            id: UUID(),
            userId: UUID(),
            provider: .gmail,
            status: .active,
            providerEmail: "user@gmail.com",
            scopes: ["https://www.googleapis.com/auth/gmail.readonly"],
            lastSyncAt: Date(),
            createdAt: Date(),
            updatedAt: Date()
        )
    }
}

File: android/app/src/main/java/net/melisma/selko/data/model/Integration.kt
package net.melisma.selko.data.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlin.time.Instant

@Serializable
enum class IntegrationProvider {
    @SerialName("gmail") GMAIL,
    @SerialName("outlook") OUTLOOK,
    @SerialName("google_photos") GOOGLE_PHOTOS,
    @SerialName("google_calendar") GOOGLE_CALENDAR
}

@Serializable
enum class IntegrationStatus {
    @SerialName("active") ACTIVE,
    @SerialName("expired") EXPIRED,
    @SerialName("revoked") REVOKED,
    @SerialName("error") ERROR
}

@Serializable
enum class IntegrationRecoveryStatus {
    @SerialName("pending") PENDING,
    @SerialName("processing") PROCESSING,
    @SerialName("waiting") WAITING,
    @SerialName("completed") COMPLETED,
    @SerialName("completed_with_errors") COMPLETED_WITH_ERRORS,
    @SerialName("failed") FAILED,
    @SerialName("superseded") SUPERSEDED
}

@Serializable
data class IntegrationRecovery(
    val id: String,
    @SerialName("integration_id") val integrationId: String,
    @SerialName("user_id") val userId: String,
    val provider: IntegrationProvider,
    val status: IntegrationRecoveryStatus,
    @SerialName("discovered_count") val discoveredCount: Int? = null,
    @SerialName("completed_count") val completedCount: Int? = null,
    @SerialName("remaining_count") val remainingCount: Int? = null,
    @SerialName("error_detail") val errorDetail: String? = null,
    @SerialName("requested_at") val requestedAt: Instant? = null
) {
    val isActive: Boolean
        get() = status == IntegrationRecoveryStatus.PENDING ||
            status == IntegrationRecoveryStatus.PROCESSING ||
            status == IntegrationRecoveryStatus.WAITING

    val needingAttentionCount: Int
        get() = maxOf(0, (discoveredCount ?: 0) - (completedCount ?: 0))
}

@Serializable
data class Integration(
    val id: String,
    @SerialName("user_id") val userId: String,
    val provider: IntegrationProvider,
    val status: IntegrationStatus,
    @SerialName("provider_email") val providerEmail: String? = null,
    val scopes: List<String> = emptyList(),
    @SerialName("last_sync_at") val lastSyncAt: Instant? = null,
    @SerialName("created_at") val createdAt: Instant? = null,
    @SerialName("updated_at") val updatedAt: Instant? = null
) {
    val isActive: Boolean
        get() = status == IntegrationStatus.ACTIVE
}
